[2단계 - 페이먼츠] 데이지(김유경) 미션 제출합니다.

kimyou1102 opened the pull request:
## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입: [미션 배포 사이트](https://kimyou1102.github.io/react-payments), [스토리북 배포 사이트](https://6800afab6301b41dd9299995-zrsmhzxgap.chromatic.com/)

### 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
이번 미션에서 상태를 어떻게 나눌지 고민을 많이 했습니다.

레벨1때 최상단에서 상태가 모두 모아져서 추후 기능 추가시 복잡해지겠다는 생각과 불필요한 렌더 및 다른 컴포넌트로의 불필요한 상태 전파가 발생했습니다. 

이번 미션도 예전과 비슷하게 최상단에 많은 상태가 모이는 문제가 발생했고 이를 해결하기 위해 **관심사 분리**와 **유지보수** 측면을 목표로 삼고 상태관리를 리팩토링 했습니다.

그래서 최상단인 App에서는 꼭 알아야하는 인풋 value에 대한 상태만 두고, CardForm컴포넌트에서 에러관련 상태들을 관리하고 싶었습니다.

#### 시도1 - 하나의 인풋섹션에 두개의 훅(상태변화 훅, 검증 훅)

상태변화에 대한 훅, 검증에 대한훅 이렇게 하나의 인풋섹션에 대해 훅을 2개 두려고 했습니다.

이렇게 분리해 보니 검증 훅에서 useEffect를 사용해야 했고 렌더 횟수가 2배가 되어서 불필요한 렌더가 많이 일어나서 이방법을 사용하지 않았습니다.

#### 시도2 - 에러 관련 상태를 useReducer로 관리

isError상태를 객체화해서 CardForm컴포넌트에서 에러 상태를 모두 다루자 하였습니다. 이렇게하면 특정 인풋의 변화마다 여러 상태를 변경해야하고, 검증 로직까지 있어 상태변화가 복잡하다 생각해서 `useReducer`를 사용하게 됐습니다.

useReducer를 적용하면서 일괄적인 상태 업데이트를 위해 `onChange` 대신 `setState`만을 전달하게 되었고, 입력 제한 및 검증 로직은 `Input` 컴포넌트에 정규식 props를 추가하여 처리하게 되었습니다. (쵸파가 저번 미션에 제안해주신 덕분에 확장성 있는 `Input` 컴포넌트를 설계하는 데 큰 도움이 되었습니다👍👍)

이렇게 하다본 온체인지에 관한 훅도 필요가 없어져서 폼에 관한 커스텀훅이 1개만 나왔습니다.

1. 미션 의도 부합 여부
  
    현재 useReducer로 상태를 관리하는 커스텀훅 1개만 있는데 이게 미션의 의도와 부합하는지 잘 모르겠습니다. 관리 로직을 분리만 한 것도 괜찮은지 혹은 재사용할 방법이 있다면 힌트 주시면 감사하겠습니다!
    
2. `useReducer` 선택의 적절성
    
    **관심사 분리**와 **유지보수** 측면에서 useReducer를 사용한 에러 상태 관리가 올바른 방향이었는지 궁금합니다. 제 목표에 부합하지 않다면 리팩토링 방향을 알려주시면 감사하겠습다!
    
3. 향후 확장성 대응
    
    카드 소유자 이름, 주소 등 새로운 입력 필드가 추가될 경우, 현재의 `useCardValidationError` 훅이 많이 커질 거 같습니다.
    그럴때는 useReducer에서 관리하는 상태를 세분화해서 나눠야할지, 어떤 식으로 리팩토링하면 좋을지 방향 알려주시면 감사하겠습니다!

#### 2) 이번 리뷰를 통해 논의하고 싶은 부분
처음에 useReducer가 관리하는 상태 안에 에러관련 메시지뿐만 아니라 show상태(각 입력 단계의 표시)도 있었습니다. 

onChange시 에러메시지와 함께 show상태도 한번에 변화하는 것이라서 한 상태에 뒀는데, 동시에 여러 상태가 변해야 할 때는 편리하나 개별 상태를 세분화하기 어려웠습니다. 

그래서 입력 유효성 검사 후에 다음 필드를 순차적으로 보여 주기 위해 UI 표시 로직을 분리하고, `useState`와 `useEffect`로 개별 상태를 관리하도록 변경했습니다. 

이 과정에서 불가피하게 한 번 더 렌더링이 발생하게 되었지만 저는 `상태를 세분화 하는 것` 이 제 목표였던 관심사 분리와 유지보수성에서 더 중요한 문제라고 판단했습니다.

쵸파는 “한 번에 여러 상태를 묶어 일괄적으로 업데이트하는 방식”과 “세분화된 상태 관리 방식” 중 어떤 접근이 더 적합하다고 보시는지 조언 부탁드립니다!


- 전 FormState 구조
    ```
    interface FormState = {
      isCardNumberError: boolean[];
      isErrorCardValidityPeriod: { month: boolean; year: boolean };
      isCardCVCError: boolean;
      isErrorCardPassword: boolean;
      cardNumberErrorMessage: string;
      cardCVCErrorMessage: string;
      cardValidityMessage: string;
      cardPasswordErrorMessage: string; 
      showRef: {
        cardCompany: false,
        validityPeriod: false,
        cvc: false,
        password: false,
      },
    };
    ```
    
- 후
    
    ```
    interface FormState {
      isCardNumberError: boolean[];
      isErrorCardValidityPeriod: { month: boolean; year: boolean };
      isCardCVCError: boolean;
      isErrorCardPassword: boolean;
      cardNumberErrorMessage: string;
      cardCVCErrorMessage: string;
      cardValidityMessage: string;
      cardPasswordErrorMessage: string;
    }
    ```
---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

wmakerjun committed:
docs: PR 템플릿 추가

kimyou1102 committed:
refactor: 카드정보 입력에 대한 input요소들의 type을 number에서 tel로 변경

- type=number일시 한글 입력이 막아지지 않는 문제 발생
- 모바일에서 숫자키패드가 보이는 tel로 변경
- 숫자가 아닌 문자가 입력될시 정규식으로 제거
- type이 number가 아니라 number에서 허용해주는 -+.eE 를 제한하는 onKeyDown이벤트 제거

kimyou1102 committed:
docs: 2단계 구현 기능 목록 작성

kimyou1102 committed:
feat: 하나의 입력에 대한 유효성 검사 통과시 다음 입력 필드 나타나게 구현

kimyou1102 committed:
docs: 구현 사항 추가 작성

kimyou1102 committed:
feat: CardCompanySelect 컴포넌트 생성

kimyou1102 committed:
refactor: 보여질 단계 변경

- 카드번호 입력 통과 후 유효기간이 아닌 카드사 선택입력이 보여지도록 변경

kimyou1102 committed:
feat: 카드사 선택을 위해 카드사선택 컴포넌트 추가

kimyou1102 committed:
feat: 카드사 선택시 미리보기 카드의 색상을 변경

kimyou1102 committed:
docs: 구현 사항 추가 작성

kimyou1102 committed:
feat: CardPasswordField 컴포넌트 생성

kimyou1102 committed:
feat: 카드 입력을 위해 CardPasswordField 컴포넌트 추가

kimyou1102 committed:
refactor: cardPasswordErrorMessage변경 시 if else문 대신 삼항연산자로 변경

kimyou1102 committed:
refactor: CardForm컴포넌트에서 에러 상태를 관리하도록 분리

- 최상단인 App컴포넌트에서 모든 상태를 관리하면 에러상태를 알지 않아도 되는 CardPreview까지 알게되어 상태를 좁게 사용하기 위해 변경
- CardForm에서 모든 인풋의 에러 상태를 관리하기 위해 객체화 한 상태를 useReducer로 관리

kimyou1102 committed:
refactor: month/year 검증 로직 중복 제거하고 validateExpiry 함수 적용

kimyou1102 committed:
refactor: 카드 유효성 검사 에러 상태를 custom 훅으로 분리

kimyou1102 committed:
refactor(CardForm): 일괄적인 상태 갱신을 위해 setState 전달로 변경하고 검증 로직 외부화

- CardForm에서는 순수 setState 및 dispatch만 수행
- Input 컴포넌트에 regexString 프로퍼티 추가
- handleChange 함수에서 정규식 테스트 후 onChange 호출
- maxLength 속성으로 길이 제한

kimyou1102 committed:
feat: autoFocus 속성을 이용해 다음 입력 단계로 넘어갈 시 input이 자동적으로 포커스 되도록 구현

kimyou1102 committed:
feat: 모두 유효한 값으로 입력되었을 때 버튼이 활성화 되도록 확인 버튼 구현

kimyou1102 committed:
chore: 라우터 설치 및 세팅

kimyou1102 committed:
feat: Home페이지 작성

- App컴포넌트의 내용 Home으로 이동

kimyou1102 committed:
chore: react-router 대신 react-router-dom으로 변경

- 브라우저에서 동작하는 useNavigate 와 같은 기능을 사용하기 위해 변경

kimyou1102 committed:
feat: 확인 페이지 구현

kimyou1102 committed:
feat: 확인 버튼 클릭시 완료 페이지로 이동 기능 구현

kimyou1102 committed:
feat: 완료페이지의 확인버튼 클릭시 다시 카드 등록 페이지로 이동하는 기능 구현

kimyou1102 committed:
style: 입력폼 css 수정

- 인풋 섹션 간격 추가
- 폼 margin-top 추가
- 폼 요소에 scroll 추가

kimyou1102 committed:
feat: CardCompanySelect 스토리북 작성

kimyou1102 committed:
feat: CardPasswordField 컴포넌트 스토리북 작성

kimyou1102 committed:
refactor: CardPreview 컴포넌트 스토리북 카드사에 따른 스토리 추가 작성

kimyou1102 committed:
refactor: 사용하지 않는 커스텀훅으로 인해 args로만 상태를 고정하여 UI를 확인하는 식으로 스토리 수정

kimyou1102 committed:
feat: 포커스 이동 커스텀훅 생성

kimyou1102 committed:
feat: 유효성 검사 통과하면 다음 인풋으로 넘어가도록 구현

- 카드넘버 입력, 유효기간 입력 처럼 인풋이 여러개인 경우 유효성 검사를 통과하면 다음 인풋으로 자동 포커스 되도록 커스텀훅 사용

kimyou1102 committed:
docs: 구현한 사항들 체크 표시

kimyou1102 committed:
fix: 현재 입력하고 있는 다음 인풋이 아닌 마지막 포커스 이동되는 버그 수정

kimyou1102 committed:
fix: 배포시 404오류 해결 위해 basename 추가

bassyu commented on src/App.tsx:
App 컴포넌트는 이제 쓰이지 않는 것 같습니다~!

kimyou1102 replied:
수정커밋: [refactor: 사용하지 않는 파일 제거](https://github.com/woowacourse/react-payments/pull/493/commits/e994c81af07a3ac3d8542e1c4ee1434bd859b37f)

bassyu commented on src/hooks/useInputFocus.tsx:
지웠을 때 자동으로 뒤로 가거나, 좌우 화살표를 눌렀을 때에도 포커싱이 이동되면 좋을 것 같아요~!

kimyou1102 replied:
그런 기능이 추가되면 UX가 더 좋아질 거 같네요!!

kimyou1102 replied:
수정커밋: [feat: useInputFocus 기능 추가](https://github.com/woowacourse/react-payments/pull/493/commits/d581bdcd2f6e50835d69c1b6f9eea86c9eafdb01)

bassyu replied:
👍 👍 훨씬 편해졌네요~!

bassyu replied:
👍 👍 훨씬 편해졌네요~!

bassyu commented on src/App.tsx:
```diff
@@ -1,83 +1,62 @@
+  cardCompany: 'cardCompany',
+  cvc: 'cvc',
+  password: 'password',
+} as const;
```
key, value가 동일하고, 앞으로도 동일할 예정으로 보이네요!
그래서 아래처럼 타입으로 관리해도 괜찮을 것 같아요~!

```suggestion
export type CARD_IFNO_INPUT_STEP = 'cardCompany' | 'cvc' | 'password';
```

bassyu commented on src/hooks/useCardValidationError.tsx:
```diff
@@ -0,0 +1,240 @@
+    });
+  };
+
+  useEffect(() => {
```
> 그래서 입력 유효성 검사 후에 다음 필드를 순차적으로 보여 주기 위해 UI 표시 로직을 분리하고, useState와 useEffect로 개별 상태를 관리하도록 변경했습니다.

렌더링을 최적화하는 것보다, 유지보수를 우선시 한 기준은 충분히 좋다고 생각합니다! 👍 👍 

다만 `useEffect`를 통한 불필요한 렌더링은 최적화를 넘어서, 버그의 위험이 높고 디버깅도 어렵기 때문에 추천드리지 않습니다.
(유지보수에 좋지 않게됩니다!)
[관련 링크](https://ko.react.dev/learn/you-might-not-need-an-effect)

아래의 로직을 `상태가 업데이트되고 난 후`가 아니라 `상태를 업데이트할 때`수행하도록 하면 될 것 같아요!

kimyou1102 replied:
최대한 useEffect를 제거하는 방향으로 리팩토링 해보겠습니다!

kimyou1102 replied:
수정커밋: [refactor: useEffect에서 show상태를 업데이트 하는 것에서 useRef 기반으로 변경](https://github.com/woowacourse/react-payments/pull/493/commits/4b747676ccc1ed7e74da7aed3c56694b09d47349)

useEffect를 제거하기 위해 show상태를 없애고 각 컴포넌트에서 ref를 기반으로 show 변수를 만들고 show상태에 따라 렌더링을 시켜주는 식으로 변경했습니다.

ref를 사용한 이유는 한번 true가 되면 false가 되지 못하도록 하기 위해 렌더링시 state가 변경되면 안되기 때문에 사용했습니다.

ref는 렌더링시 이용하면 안된다는 공식문서의 글을 보고 이런식으로 변수화해서 사용했는데 이렇게 사용한적이 처음이라 올바른 방향인지 잘모르겠네요. 쵸파가 보기에는 이러한 리팩토링 방향이 괜찮아 보이시나요??

bassyu replied:
useEffect 제거해주신 것 좋네요! 👍  다만 ref 는 다른 해결책이 보이네요!

show 상태에 따라서 보여줄지 말지 판단하는 것이기 때문에, show 상태는 UI에 영향을 주는 데이터, 즉 관리되어야 하는 state 중 하나입니다.
한번 true가 된 이후로 false가 되지 않길 원한다면, 부모 컴포넌트가 각 필드의 show 상태를 가지고 있고, 컨트롤 하는 것이 좋습니다!

kimyou1102 replied:
수정커밋: [refactor: 불필요한 ref대신 state로 변경](https://github.com/woowacourse/react-payments/pull/493/commits/5103709aea428866f4f7d7e9e1019d58b2424318)

bassyu commented on src/hooks/useCardValidationError.tsx:
> 쵸파는 “한 번에 여러 상태를 묶어 일괄적으로 업데이트하는 방식”과 “세분화된 상태 관리 방식” 중 어떤 접근이 더 적합하다고 보시는지 조언 부탁드립니다!

질문을 요약해보자면, 지금 상황에서 상태를 그룹화 할지 말지 가 맞을까요?

정답은 없지만, 저는 `입력값에 관련된 UI`랑, `다음 입력 필드를 보여줄지 말지`를 각각 그룹화해서 관리할 것 같습니다!

bassyu commented on src/hooks/useCardValidationError.tsx:
```diff
@@ -0,0 +1,240 @@
+  cardValidityPeriod,
+  showNextStep,
+}: useCardValidationErrorProps) {
+  const [state, dispatch] = useReducer(formReducer, initialState);
```
> useReducer 선택의 적절성

파편화된 상태가 있었고, 이들을 그룹화 할 필요가 있었고, 객체로 그룹화를 해보니, reducer가 재격이네?
이런 흐름이었다면 적절한 것 같습니다! 그래서 도구 사용만 봤을 때에는 잘 사용하신 것 같아요 👍 👍 

중요한 것은, useReducer의 사용 자체 보다는, 파편화된 여러 상태를 **무엇을 기준으로 그룹화 했는지**라고 생각합니다~!

예를 들어서 지금 구조는, error와 errorMessage는 그룹화 되어있지만, value는 조금 먼 곳에 정의되어 있는데요.
새로운 입력 필드가 생기면, value와 error의 거리가 좀 멀어서 아쉬울수도 있겠네용

[다른 예시](https://github.com/woowacourse/react-payments/pull/457#discussion_r2061213901)도 참고해보면서 고민해보면 좋을 것 같습니다!

kimyou1102 replied:
error관련 상태가 CardForm에서만 사용되어서 내부에 유지하는 것이 유지보수관점에서 더 좋다는 생각으로 이런 구조를 만들었습니다.

쵸파의 제안처럼 input 상태와 error가 합쳐지면 상태를 세분화할 수 없지 않을까? 하는 고민이 됩니다. 

한곳에 있다면 입력 필드가 추가되었을때 한번에 추가할 수 있어서 좋겠지만 상태가 세분화되지 못해서 각각의 컴포넌트만 알 수 있게 한다던가 하는 것은 하지 못할 거 같습니다. 

보통 이런 경우 한곳에서 인풋에 관한 정보들을 관리하는 것이 유지보수성이 더 좋은건지 궁금합니다!!

bassyu replied:
좋은 포인트 잡아주셨네요! 질문 주셔서 감사합니다 👍 
데이지 말대로, 값은 필요하지만 에러는 필요없는 컴포넌트가 있을 수도 있는데, 이럴 때 세분화가 안되는 단점이 있죠.

개인적으로는, 그래도 form의 어떤 값이랑 에러는 워낙 관련이 높기 때문에, 같이 만들어지고 같이 이동하는 안전성의 이점이 크다고 생각합니다~!

bassyu commented on src/components/cardInfoForm/CardForm/CardForm.tsx:
```diff
@@ -0,0 +1,240 @@
+    cardNumber,
+    cardCVC,
+    cardValidityPeriod,
+    showNextStep,
```
네이밍을 이렇게 하면, callback의 실행 시점을 짐작할 수 있는 장점이 있습니다~!

```suggestion
    onChangeStep: showNextStep,
```

bassyu commented on src/components/cardInfoForm/CardPasswordField/CardPasswordField.tsx:
```diff
@@ -0,0 +1,44 @@
+          isError={isError}
+          type="password"
+          name="cardPassword"
+          id="cardPassword"
```
`CardPasswordField` 컴포넌트가 재사용되면, id가 중복되는 이슈가 발생하겠네요~!
어떻게 해결해 볼 수 있을까요?

kimyou1102 replied:
1단계에서도 이부분에 대해 피드백을 받았는데 구현하다 보니 또 놓쳤네요😢 useId훅을 이용해 고유한 값을 만들 거 같습니다!

kimyou1102 replied:
수정커밋: [refactor: useId를 활용해 고유한 id로 변경](https://github.com/woowacourse/react-payments/pull/493/commits/1ea3f5b0287620379fab606a6d9c8f12657aae37)

bassyu commented on src/components/cardInfoForm/CardPasswordField/CardPasswordField.tsx:
```diff
@@ -0,0 +1,44 @@
+
+export default CardPasswordField;
+
+const InputWrapper = styled.div`
+  display: flex;
+  gap: 10px;
+  margin-top: 8px;
+`;
+const Label = styled.label`
+  font-size: 12px;
+`;
```
꽤 중복이 있어서, 중복 제거 해볼만한 것 같습니다!

kimyou1102 replied:
수정커밋: [refactor: LabeledInput 컴포넌트 이용해서 중복되는 스타일 컴포넌트 제거](https://github.com/woowacourse/react-payments/pull/493/commits/9d2def15af1d994d61f4396fdadb32642cfda9d5)

bassyu reviewed (changes requested):
안녕하세요 데이지~!
2단계도 잘 부탁드립니다!!

우선 기능을 봤을 때, 검증 메세지가 자연스럽고 적절하게 나와서 좋네요 💯 💯 

관심사와 유지보수를 고민하는 자세도 좋다고 생각합니다~!
훅 위주로 리팩터링 해봐도 좋을 것 같아요.

논의할 점이나 질문 답변은 코멘트로 남겨두겠습니다!
2단계도 수고 많으셨습니다! 👍 👍

kimyou1102 committed:
refactor: useEffect에서 show상태를 업데이트 하는 것에서 useRef 기반으로 변경

- 각 컴포넌트에서 useRef를 활용해 한 번 true가 되면 유지되는 show 변수 구현
- 이를 위해 CardCVCSection, CardCompanySection, CardPasswordSection 컴포넌트 구현
- 불필요한 useEffect 제거

kimyou1102 committed:
refactor: 사용하지 않는 파일 제거

kimyou1102 committed:
refactor: useId를 활용해 고유한 id로 변경

kimyou1102 committed:
feat: useInputFocus 기능 추가

- 인풋의 내용을 모두 지우면 이전 인풋으로 포커스되는 기능 추가
- left, rigth 키보드 동작으로 이전, 이후 인풋으로 포커스되는 기능 추가

kimyou1102 committed:
feat: LabeledInput 컴포넌트 생성

- inputWrapper 와 label이 있는 반복되는 코드를 줄이기 위해 생성

kimyou1102 committed:
refactor: LabeledInput 컴포넌트 이용해서 중복되는 스타일 컴포넌트 제거

kimyou1102 committed:
refactor: 폴더 구조 변경

- cardInfo폴더의 컴포넌트들이 많아져서 가독성을 위해 관심사별 하위폴더 생성

kimyou1102 committed:
fix: isValid 함수의 반환 타입을 수정

kimyou1102 committed:
fix: 카드숫자 입력의 인풋이 비었을때 error상태가 아닌 오류 수정

kimyou1102 commented:
안녕하세요 쵸파! 미션을 진행하다보니 반영이 늦어진점 죄송합니다...ㅠㅠ
제가 고민했던 부분들에 쵸파가 답변해주셔서 많은 도움이 되었습니다!!

리뷰에 대해 이야기를 더 나누고 싶은 점들과 반영사항은 대댓글로 수정커밋을 남겨뒀습니다! (수정하면서 제거하게된 부분들에 대한 수정사항은 제외했습니다.)

bassyu reviewed (changes requested):
안녕하세요 데이지~!
코멘트 잘 이해하시고 다시 질문 주셔서 좋네요 👍 👍 

데이지가 질문 주신 `ref` 부분은 개선이 필요해 보여서 다시 코멘트 남겼습니다!
이 부분 해결되면 이번 미션은 마무리해도 좋을 것 같아요~!

bassyu commented on src/components/cardInfoForm/cardInfoInputSections/CardPasswordSection/CardPasswordSection.tsx:
```diff
@@ -0,0 +1,46 @@
+  errorMessage,
+  isValid,
+}: CardPasswordSectionProps) {
+  const showRef = useRef(false);
+
+  if (!showRef.current && isValid()) {
+    showRef.current = true;
+  }
+  const show = showRef.current;
```
이것도 제가 남긴 거랑 비슷한 코멘트네요!

kimyou1102 committed:
refactor: 불필요한 ref대신 state로 변경

- stepShow상태를 onChange에서 변경시켜주기 위해 value, isError, errorMessage를 하나의 객체 상태로 관리하도록 변경
- ref제거

kimyou1102 commented:
안녕하세요 쵸파! ref를 없애기 위해 구조를 변경했습니다.

ref를 없애기 위해 부모에서 show상태를 두려고 하니 onChange에서 한꺼번에 변경해줘야 했어서 결국 value, isError, errorMessage를 하나의 상태로 관리하도록 변경됐습니다.

이번 미션에서 쵸파 덕분에 상태에 대해 더 많이 고민해볼 수 있어서 좋았습니다. 감사합니다😊

bassyu reviewed (approved):
안녕하세요 데이지!

어제 확인을 했는데, Approve를 안했네요 😓 
ref 대신 상태관리 활용해 주시면서, 함께 리팩터링 잘 해주셨네요! 👍 👍 

이번 미션은 이만 머지하겠습니다~! 정말 수고 많으셨습니다!
다음 미션도 파이팅입니다!!
